pallets/oracle/src/dia.rs

```diff
@@ -0,0 +1,114 @@
+						return Some((
+							DOT_DIA_BLOCKCHAIN.as_bytes().to_vec(),
+							DOT_DIA_SYMBOL.as_bytes().to_vec(),
+						)),
+					primitives::ForeignCurrencyId::KSM =>
+						return Some((
+							KSM_DIA_BLOCKCHAIN.as_bytes().to_vec(),
+							KSM_DIA_SYMBOL.as_bytes().to_vec(),
+						)),
+					_ => unimplemented!(),
+				},
+				CurrencyId::Native => unimplemented!(),
+				CurrencyId::StellarNative => unimplemented!(),
+				CurrencyId::AlphaNum4 { .. } => unimplemented!(),
+				CurrencyId::AlphaNum12 { .. } => unimplemented!(),
+			},
+			OracleKey::FeeEstimation => Some((vec![6u8], vec![])),
+		}
+	}
+}
+
+impl Convert<(Vec<u8>, Vec<u8>), Option<OracleKey>> for DiaOracleKeyConvertor {
+	fn convert(dia_oracle_key: (Vec<u8>, Vec<u8>)) -> Option<OracleKey> {
+		let (blockchain, symbol) = dia_oracle_key;
+		let blockchain = String::from_utf8(blockchain);
+		let symbol = String::from_utf8(symbol);
+		match (blockchain, symbol) {
+			(Ok(blockchain), Ok(symbol)) => {
+				if blockchain == DOT_DIA_BLOCKCHAIN && symbol == DOT_DIA_SYMBOL {
+					return Some(OracleKey::ExchangeRate(CurrencyId::XCM(
+						primitives::ForeignCurrencyId::DOT,
+					)))
+				} else if blockchain == KSM_DIA_BLOCKCHAIN && symbol == KSM_DIA_SYMBOL {
+					return Some(OracleKey::ExchangeRate(CurrencyId::XCM(
+						primitives::ForeignCurrencyId::KSM,
+					)))
+				} else {
+					return None
+				}
+			},
+			(_, _) => return None,
+		}
+	}
+}
+
+pub struct DiaOracleAdapter<
+	DiaPallet: DiaOracle,
+	UnsignedFixedPoint,
+	Moment,
+	ConvertKey,
+	ConvertPrice,
+	ConvertMoment,
+>(
+	marker::PhantomData<(
+		DiaPallet,
+		UnsignedFixedPoint,
+		Moment,
+		ConvertKey,
+		ConvertPrice,
+		ConvertMoment,
+	)>,
+);
+
+impl<Dia, UnsignedFixedPoint, Moment, ConvertKey, ConvertPrice, ConvertMoment>
+	DataProviderExtended<OracleKey, TimestampedValue<UnsignedFixedPoint, Moment>>
+	for DiaOracleAdapter<Dia, UnsignedFixedPoint, Moment, ConvertKey, ConvertPrice, ConvertMoment>
+where
+	Dia: DiaOracle,
+	ConvertKey: Convert<OracleKey, Option<(Vec<u8>, Vec<u8>)>>
+		+ Convert<(Vec<u8>, Vec<u8>), Option<OracleKey>>,
+	ConvertPrice: Convert<u128, Option<UnsignedFixedPoint>>,
+	ConvertMoment: Convert<u64, Option<Moment>>,
+{
+	fn get_no_op(key: &OracleKey) -> Option<TimestampedValue<UnsignedFixedPoint, Moment>> {
+		let (blockchain, symbol) = ConvertKey::convert(key.clone())?;
+
+		let Ok(coin_info) = Dia::get_coin_info(blockchain, symbol) else {
+            return None;
+        };
+
+		let value = ConvertPrice::convert(coin_info.price)?;
+		let Some(timestamp) = ConvertMoment::convert(coin_info.last_update_timestamp) else{
+            return None;
+        };
+
+		Some(TimestampedValue { value, timestamp })
+	}
+
+	///do not need this funtion implementation
+	fn get_all_values() -> Vec<(OracleKey, Option<TimestampedValue<UnsignedFixedPoint, Moment>>)> {
+		panic!("spacewalk oracle extension does not requre implementation of DataProviderExtended get_all_values function")
+	}
```

Comment:
Thanks! You are right. 
But the problem is that we even can't implement this function for DIA oracle. Because they did not provide the function in `DiaOracle` trait that will return list of key or all values. This is limitation of DIA oracle pallet.
here is like DIAOracle trait looks like
```
pub trait DiaOracle {
    /// Returns the coin info by given name
    fn get_coin_info(blockchain: Vec<u8>, symbol: Vec<u8>) -> Result<CoinInfo, DispatchError>;

    /// Returns the price by given name
    fn get_value(blockchain: Vec<u8>, symbol: Vec<u8>) -> Result<PriceInfo, DispatchError>;
}
```

Reply:
Hmm, I see. But we can query the coin info of all items by using the `Dia::prices_map()` getter function (see [here](https://github.com/pendulum-chain/oracle-pallet/blob/af8517a2d9eb469c9681195e6653aa519f05cf28/pallets/dia-oracle/src/lib.rs#L109-L111)). So I guess it actually should be possible?

Reply:
> Hmm, I see. But we can query the coin info of all items by using the `Dia::prices_map()` getter function (see [here](https://github.com/pendulum-chain/oracle-pallet/blob/af8517a2d9eb469c9681195e6653aa519f05cf28/pallets/dia-oracle/src/lib.rs#L109-L111)). So I guess it actually should be possible?

Thanks. You are right. But. this function not included to DIAOracle trait that they provide.
We use bound 
```
impl<Dia, UnsignedFixedPoint, Moment, ConvertKey, ConvertPrice, ConvertMoment>
	DataProviderExtended<OracleKey, TimestampedValue<UnsignedFixedPoint, Moment>>
	for DiaOracleAdapter<Dia, UnsignedFixedPoint, Moment, ConvertKey, ConvertPrice, ConvertMoment>
where
	Dia: DiaOracle,
	ConvertKey: Convert<OracleKey, Option<(Vec<u8>, Vec<u8>)>>
		+ Convert<(Vec<u8>, Vec<u8>), Option<OracleKey>>,
	ConvertPrice: Convert<u128, Option<UnsignedFixedPoint>>,
	ConvertMoment: Convert<u64, Option<Moment>>,
{}
```
**Dia: DiaOracle**
Of course we can add more not flexible logic to retrieve data about all keys from DIA oracle but actually we **do not need it**. Because we store all required keys for spacewalk oracle in storage `OracleKeys<Vec<OracleKey>>`.